libia2/threads.c

```diff
@@ -0,0 +1,85 @@
+#include <pthread.h>
+
+#include "ia2.h"
+
+extern void init_stacks(void);
+
+struct ia2_thread_thunk {
+  void *(*fn)(void *);
+  void *data;
+};
+
+void *ia2_thread_begin(void *arg) {
+  struct ia2_thread_thunk *thunk = (struct ia2_thread_thunk *)arg;
+  void *(*fn)(void *) = thunk->fn;
+  void *data = thunk->data;
+  munmap(arg, sizeof(struct ia2_thread_thunk));
+
+  init_stacks();
+
+  // TODO: switch to compartment stack before calling `fn`
+
+  return fn(data);
+}
+
+extern int __real_pthread_create(pthread_t *restrict thread,
+                                 const pthread_attr_t *restrict attr,
+                                 void *(*fn)(), void *data);
+
+int __wrap_pthread_create(pthread_t *restrict thread,
+                          const pthread_attr_t *restrict attr, void *(*fn)(),
+                          void *data) {
+  void *mmap_res = mmap(NULL, sizeof(struct ia2_thread_thunk),
+                        PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON, -1, 0);
+  if (mmap_res == MAP_FAILED) {
+    return -1;
+  }
+  struct ia2_thread_thunk *thread_thunk = (struct ia2_thread_thunk *)mmap_res;
+  thread_thunk->fn = fn;
+  thread_thunk->data = data;
+  return __real_pthread_create(thread, attr, ia2_thread_begin, thread_thunk);
+}
+
+/* guards `thread_exists_bitmap` */
+static pthread_mutex_t thread_exists_bitmap_lock IA2_SHARED_DATA =
+    PTHREAD_MUTEX_INITIALIZER;
+/* tracks which compact tid values are used */
+static bool thread_exists_bitmap[IA2_MAX_THREADS] IA2_SHARED_DATA = {0};
+
+size_t ia2_alloc_compacted_tid(void) {
+  pthread_mutex_lock(&thread_exists_bitmap_lock);
+  for (size_t i = 0; i < IA2_MAX_THREADS; i++) {
+    if (thread_exists_bitmap[i] == 0) {
+      thread_exists_bitmap[i] = 1;
+      pthread_mutex_unlock(&thread_exists_bitmap_lock);
+      return i;
+    }
+  }
+  printf("internal error: too many threads\n");
+  exit(1);
+}
+
+__thread size_t ia2_compacted_tid;
+
+size_t ia2_get_compacted_tid(void) { return ia2_compacted_tid; }
+
+_Noreturn extern void __real_pthread_exit(void *retval);
+
+_Noreturn void ia2_release_own_compacted_tid(void *retval) {
+  thread_exists_bitmap[ia2_compacted_tid] = 0;
```

Comment:
Don't we need `pthread_mutex_lock` here?

Reply:
I think you're right, we do. I had thought we might be safe because there's only one concurrent writer to this index, since obtaining ownership of that index with `ia2_alloc_compated_tid` is locked, but that relies on `ia2_compacted_tid` always being correct, and still means we have synchronization around readers and (setting) writers but not around clearing writers... there's no reason not to lock here.

Reply:
Fixed.